File: tests/test_incremental.py
"""Tests for ldoce5viewer.incremental (Maker / Searcher)."""


import pytest

from ldoce5viewer.incremental import IndexError as IncrementalIndexError
from ldoce5viewer.incremental import Maker, Searcher


def _build_index(tmp_path, items):
    """Build an incremental index from *items* and return the index path.

    Each item is a tuple: (plain, typecode, label, path, prio).
    """
    index_path = str(tmp_path / "inc.dat")
    tmp_path2 = str(tmp_path / "inc.tmp")
    maker = Maker(index_path, tmp_path2)
    for plain, typecode, label, p, prio in items:
        maker.add_item(plain, typecode, label, p, prio)
    maker.finalize()
    return index_path


ITEMS = [
    ("apple", "hw", "Apple", "entry/apple", 0),
    ("application", "hw", "Application", "entry/application", 0),
    ("apply", "hw", "Apply", "entry/apply", 1),
    ("banana", "hw", "Banana", "entry/banana", 0),
    ("band", "hw", "Band", "entry/band", 0),
]


class TestIncrementalRoundTrip:
    def test_exact_prefix_match(self, tmp_path):
        index_path = _build_index(tmp_path, ITEMS)
        with Searcher(index_path) as searcher:
            results = searcher.search("appl", limit=10)
        labels = [r[0] for r in results]
        assert "Apple" in labels
        assert "Application" in labels
        assert "Apply" in labels
        # "banana" should not appear
        assert "Banana" not in labels

    def test_full_word_match(self, tmp_path):
        index_path = _build_index(tmp_path, ITEMS)
        with Searcher(index_path) as searcher:
            results = searcher.search("banana", limit=10)
        assert len(results) >= 1
        assert results[0][0] == "Banana"

    def test_no_match_returns_empty(self, tmp_path):
        index_path = _build_index(tmp_path, ITEMS)
        with Searcher(index_path) as searcher:
            results = searcher.search("zzz", limit=10)
        assert results == []

    def test_limit_is_respected(self, tmp_path):
        index_path = _build_index(tmp_path, ITEMS)
        with Searcher(index_path) as searcher:
            results = searcher.search("a", limit=2)
        assert len(results) <= 2

    def test_empty_key_returns_empty(self, tmp_path):
        index_path = _build_index(tmp_path, ITEMS)
        with Searcher(index_path) as searcher:
            results = searcher.search("", limit=10)
        assert results == []

    def test_broken_file_raises(self, tmp_path):
        bad_path = str(tmp_path / "bad.dat")
        (tmp_path / "bad.dat").write_bytes(b"\x00" * 4)
        with pytest.raises(IncrementalIndexError):
            Searcher(bad_path)

    def test_result_structure(self, tmp_path):
        """Each result is (label, path, plain, prio, None)."""
        index_path = _build_index(tmp_path, ITEMS)
        with Searcher(index_path) as searcher:
            results = searcher.search("band", limit=5)
        assert len(results) >= 1
        label, path, plain, prio, extra = results[0]
        assert isinstance(label, str)
        assert isinstance(path, str)
        assert isinstance(plain, str)
        assert isinstance(prio, int)
        assert extra is None

File: ldoce5viewer-tui/tests/test_incremental.rs
//! Integration tests for the incremental (prefix-search) index.

use std::path::Path;
use tempfile::tempdir;

use ldoce5viewer_tui::search::incremental::{IncrementalMaker, IncrementalSearcher, normalize_index_key};

// --------------------------------------------------------------------------
// Helper
// --------------------------------------------------------------------------

fn build_index(dir: &Path, items: &[(&str, &str, &str, &str, u8)]) -> std::path::PathBuf {
    let idx_path = dir.join("inc.dat");
    let tmp_path = dir.join("inc.tmp");
    let mut maker = IncrementalMaker::new(&idx_path, &tmp_path).unwrap();
    for &(plain, tc, label, path, prio) in items {
        maker.add_item(plain, tc, label, path, prio).unwrap();
    }
    maker.finalize().unwrap();
    idx_path
}

// --------------------------------------------------------------------------
// Tests
// --------------------------------------------------------------------------

#[test]
fn test_round_trip_prefix_match() {
    let dir = tempdir().unwrap();
    let idx = build_index(dir.path(), &[
        ("apple",       "hw", "Apple",       "/fs/apple",       0),
        ("application", "hw", "Application", "/fs/application", 0),
        ("apply",       "hw", "Apply",       "/fs/apply",       1),
        ("banana",      "hw", "Banana",      "/fs/banana",      0),
    ]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    let results = searcher.search("appl", 100);
    let labels: Vec<&str> = results.iter().map(|r| r.label.as_str()).collect();
    assert!(labels.contains(&"Apple"),       "Apple should match 'appl'");
    assert!(labels.contains(&"Application"), "Application should match 'appl'");
    assert!(labels.contains(&"Apply"),       "Apply should match 'appl'");
    assert!(!labels.contains(&"Banana"),     "Banana should NOT match 'appl'");
}

#[test]
fn test_exact_match() {
    let dir = tempdir().unwrap();
    let idx = build_index(dir.path(), &[
        ("run",  "hw", "Run",  "/fs/run",  0),
        ("runs", "hw", "Runs", "/fs/runs", 0),
    ]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    let results = searcher.search("run", 100);
    let labels: Vec<&str> = results.iter().map(|r| r.label.as_str()).collect();
    assert!(labels.contains(&"Run"),  "exact match 'run' missing");
    assert!(labels.contains(&"Runs"), "'runs' is a prefix match of 'run' and should be included");
}

#[test]
fn test_no_match() {
    let dir = tempdir().unwrap();
    let idx = build_index(dir.path(), &[("elephant", "hw", "Elephant", "/fs/e", 0)]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    assert!(searcher.search("xyz", 100).is_empty());
}

#[test]
fn test_empty_query() {
    let dir = tempdir().unwrap();
    let idx = build_index(dir.path(), &[("word", "hw", "Word", "/fs/word", 0)]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    assert!(searcher.search("", 100).is_empty());
}

#[test]
fn test_limit_respected() {
    let dir = tempdir().unwrap();
    let items: Vec<(String, String, String, String, u8)> = (0u32..50)
        .map(|i| (format!("word{i:03}"), "hw".to_string(), format!("Word{i}"), format!("/fs/w{i}"), 0))
        .collect();
    let refs: Vec<(&str, &str, &str, &str, u8)> = items
        .iter()
        .map(|(a, b, c, d, e)| (a.as_str(), b.as_str(), c.as_str(), d.as_str(), *e))
        .collect();
    let idx = build_index(dir.path(), &refs);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    let results = searcher.search("word", 10);
    assert_eq!(results.len(), 10, "limit should be respected");
}

#[test]
fn test_accent_normalisation() {
    let dir = tempdir().unwrap();
    let idx = build_index(dir.path(), &[("café", "hw", "Café", "/fs/cafe", 0)]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    // Searching for "cafe" should match the accented entry because both
    // are normalised to "cafe"
    let results = searcher.search("cafe", 10);
    assert!(!results.is_empty(), "normalised 'cafe' should match 'café'");
}

#[test]
fn test_sort_order_prio() {
    let dir = tempdir().unwrap();
    // Same plain text, different priority
    let idx = build_index(dir.path(), &[
        ("run", "hw", "Run (prio=2)", "/fs/run2", 2),
        ("run", "hw", "Run (prio=0)", "/fs/run0", 0),
        ("run", "hw", "Run (prio=1)", "/fs/run1", 1),
    ]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    let results = searcher.search("run", 10);
    // Should be sorted: prio=0 first, then 1, then 2
    let prios: Vec<u8> = results.iter().map(|r| r.prio).collect();
    for w in prios.windows(2) {
        assert!(w[0] <= w[1], "results not sorted by prio: {prios:?}");
    }
}

#[test]
fn test_result_fields() {
    let dir = tempdir().unwrap();
    let idx = build_index(dir.path(), &[
        ("example", "hw", "Example Label", "/dict/example", 3),
    ]);
    let searcher = IncrementalSearcher::open(&idx).unwrap();
    let results = searcher.search("example", 1);
    assert_eq!(results.len(), 1);
    let r = &results[0];
    assert_eq!(r.label, "Example Label");
    assert_eq!(r.path, "/dict/example");
    assert_eq!(r.prio, 3);
    assert_eq!(r.typecode, "hw");
}

#[test]
fn test_normalize_index_key() {
    assert_eq!(normalize_index_key("Hello"),    "hello");
    assert_eq!(normalize_index_key("  café  "), "cafe");
    assert_eq!(normalize_index_key(""),          "");
    assert_eq!(normalize_index_key("über"),      "uber");
    // Copyright sign maps to 'c'
    assert_eq!(normalize_index_key("\u{00a9}"),  "c");
    // Numbers preserved
    assert_eq!(normalize_index_key("mp3"),       "mp3");
}
